bftengine/src/communication/PlainTcpCommunication.cpp

```diff
@@ -715,100 +629,94 @@ class PlainTCPCommunication::PlainTcpImpl {
+        0,
+        _selfId,
+        ConnType::Incoming,
+        _logger,
+        _statusCallback,
+        nodes);
+    _pAcceptor->async_accept(
+        conn->socket, boost::bind(&PlainTcpImpl::on_accept, this, conn, nodes, boost::asio::placeholders::error));
     LOG_TRACE(_logger, "exit, node: " << _selfId);
   }
 
-  PlainTcpImpl(const PlainTcpImpl &) = delete; // non construction-copyable
-  PlainTcpImpl(const PlainTcpImpl &&) = delete; // non movable
-  PlainTcpImpl &operator=(const PlainTcpImpl &) = delete; // non copyable
+  PlainTcpImpl(const PlainTcpImpl &) = delete;             // non construction-copyable
+  PlainTcpImpl(const PlainTcpImpl &&) = delete;            // non movable
+  PlainTcpImpl &operator=(const PlainTcpImpl &) = delete;  // non copyable
   PlainTcpImpl() = delete;
 
   PlainTcpImpl(NodeNum selfNodeId,
                NodeMap nodes,
                uint32_t bufferLength,
                uint16_t listenPort,
                uint32_t maxServerId,
-               string listenIp,
-               UPDATE_CONNECTIVITY_FN statusCallback) :
-      _selfId{selfNodeId},
-      _listenPort{listenPort},
-      _listenIp{listenIp},
-      _bufferLength{bufferLength},
-      _maxServerId{maxServerId},
-      _statusCallback{statusCallback} {
+               string listenHost,
+               UPDATE_CONNECTIVITY_FN statusCallback)
+      : _selfId{selfNodeId},
+        _listenPort{listenPort},
+        _listenHost{listenHost},
+        _bufferLength{bufferLength},
+        _maxServerId{maxServerId},
+        _statusCallback{statusCallback} {
     // all replicas are in listen mode
     if (_selfId <= _maxServerId) {
-      LOG_DEBUG(_logger, "node " << _selfId << " listening on " << _listenPort);
-      tcp::endpoint ep(address::from_string(_listenIp), _listenPort);
-      _pAcceptor = boost::make_unique<tcp::acceptor>(_service, ep);
-      start_accept(nodes);
-    } else // clients dont need to listen
+      asio::ip::tcp::resolver::query query(asio::ip::tcp::v4(), _listenHost, std::to_string(_listenPort));
+      asio::ip::tcp::resolver resolver(_service);
+      // This throws an exception if it can't resolve, but since this is the
+      // *listen* port, we should expect to be able to resolve our own name, or
+      // shutdown otherwise.
+      boost::system::error_code ec;
+      asio::ip::tcp::resolver::iterator results = resolver.resolve(query, ec);
+      if (!ec && results != asio::ip::tcp::resolver::iterator()) {
+        asio::ip::tcp::endpoint ep = *results;
+        LOG_INFO(_logger, "Resolved " << _listenHost << ":" << _listenPort << " to " << ep);
+        _pAcceptor = boost::make_unique<tcp::acceptor>(_service, ep);
+        start_accept(nodes);
+      } else {
+        LOG_WARN(_logger, "Unable to resolve listen host (" << _listenHost << ") for node " << _selfId << ": " << ec);
+        // This node is dead in the water if it can't resolve its own listen
+        // host/ip.
+        return -1;  // failed
```

Comment:
Self review: this is a copy-paste-o from an earlier version of `TcpTlsImpl::Start`. The comment at the start is wrong - that version of `resolve` does not throw an exception. Since this is in the constructor here, it also can't return -1.

I feel like reducing duplication between the plain TCP and TLS/TCP implementations would help prevent this sort of mistake in the future. That may require changing the build to allow building both at the same time.

Reply:
Fixed in bb0343a.